fix: added keyword substitution support in edx_ace based bulk course emails

ziafazal opened the pull request:
This PR has changes to support keyword substitution in edx_ace based bulk course emails. This issue is reported [here](https://discuss.openedx.org/t/keyword-support-in-course-bulk-emails-with-edx-ace/8339/4).

Testing Instructions.
1. Login as a user having django admin access and enable bulk course email flag
<img width="588" alt="Screen Shot 2022-09-27 at 6 36 32 PM" src="https://user-images.githubusercontent.com/434995/192541599-f1046e6e-a141-4d77-91d8-66ba8e7a1109.png">
 
2. Login as course instructor or staff
3. Open `Email` tab on instructor dashboard of any course
4. Set Myself in `Send To` field, enter anything in `Subject` field. In `Message` field enter the text below and hit `Send Mail` button
```
Hey %%USER_FULLNAME%%

welcome in %%COURSE_DISPLAY_NAME%%
```
5. You should receive email with `%%USER_FULLNAME%%` and `%%COURSE_DISPLAY_NAME%%` replaced with user's full name and the course display name.

ziafazal committed:
fix: added keyword substitution support in edx_ace based bulk course email

fix: pylint quality violation

fix: added positive assert and docs link

fix: mark string for i18n according to best practices

sarina commented:
Should anything be added in the UI to show users that this capability exists?

I assume preview is difficult/impossible (it was when I previously worked on bulk email in 2013), but that would be a good consideration for extension to validate that the variables are properly entered

hurtstotouchfire commented:
Hi @ziafazal this feature is in a state of ownership limbo at 2U. Most recently my team worked on it but we will be transferring ownership to @jristau1984's team Infinity. My team also built a replacement for the front end which I don't think is fully supported in Open edX yet, and since we are transititioning ownership we don't have a current plan to address that. It might help to have a synchronous conversation about how to land this.

sarina commented on lms/djangoapps/bulk_email/tests/test_email.py:
```diff
@@ -229,6 +229,29 @@ def test_ace_eanbled_toggle(self, ace_enabled, email_sent_with_ace, mock_ace_ema
+        self.assertNotIn('%%USER_FULLNAME%%', text_message_body)
+        self.assertNotIn('%%COURSE_DISPLAY_NAME%%', text_message_body)
+        self.assertNotIn('%%USER_FULLNAME%%', html_message_body)
+        self.assertNotIn('%%COURSE_DISPLAY_NAME%%', html_message_body)
```
is it worth adding positive asserts as well? (ie in addition to testing that the variable is _not_ present, that the correct data _is_ present?)

ziafazal commented:
@sarina thanks for reviewing. I have 

- added positive asserts in test
- added some text and docs link in UI if instructors need to know more about message `Keywords`

sarina commented:
@ziafazal I want to be clear that I don't know this code area well and don't own it, so am not 100% confident to provide review. From my perspective it looks good (thanks for the updates) but I'd coordinate with @hurtstotouchfire to land it.

sarina commented on lms/templates/instructor/instructor_dashboard_2/send_email.html:
```diff
@@ -91,7 +91,10 @@
-        <p class="copy">${_("We recommend sending learners no more than one email message per week. Before you send your email, review the text carefully and send it to yourself first, so that you can preview the formatting and make sure embedded images and links work correctly.")}</p>
+        <p class="copy">
+            ${_("We recommend sending learners no more than one email message per week. Before you send your email, review the text carefully and send it to yourself first, so that you can preview the formatting and make sure embedded images and links work correctly.")}
+            ${_("You can include keywords in your messages. For list of available keywords see ")} <a href="${settings.COURSE_BULK_EMAIL_HELP_URL}" target="_blank" rel="noopener">${_("documentation")}</a>.
```
ah one last comment - this isn't correct for internationalization.
```suggestion
            ${_("You can include keywords in your messages. For list of available keywords see the {html_start}documentation{html_end}").format(html_start='<a href="${settings.COURSE_BULK_EMAIL_HELP_URL}" target="_blank" rel="noopener">', html_end='</a>')}
```

I think that works but should be tested. We can't concatenate sentences, that may break languages with different orderings and _always_ breaks RTL languages.

nedbat commented:
📣 💥 Heads-up: You must either **rebase onto master** or **merge master into your branch** to keep passing required checks.

We added a new required check, "Tests Successful," that this PR does not yet run.  Rebasing will get it started.

If you have any questions, please reach out to the Architecture team (either [#architecture](https://openedx.slack.com/archives/C024JPE27PV) on Open edX Slack or [#external-architecture](https://edx-internal.slack.com/archives/C024JPE27PV) on edX internal Slack).

e0d commented:
@hurtstotouchfire @jristau1984 There were a couple salient notes above:

- This is transitioning
- That a synchronous convo might be required to land this.

This looks valuable and straight-forward, but y'all are the most intensive users of this feature I believe.  What are the best next steps?

justinhynes reviewed (approved):
This looks good to me! 👍 

FWIW, I did do a bit of testing against this PR, but performed this testing using the new Communications MFE (which replaces the legacy email client/tab of the Instructor Dashboard). I did spot-check the documentation link (which looks good), but figured I should mention I didn't spent a ton of time messing with the legacy email feature.
